Antithesis blogpost (#1051)
etcd-io/website: content/en/blog/2025/autonomus_testing_with_antithesis.md

---
title: Autonomous Testing of etcd's Robustness

> author: "Marek Siarkowicz (Google)"

Reviewer: @serathius could we put "Marcus Hodgson (Antithesis)" here?

draft: true
---

As a critical component of many production systems, including Kubernetes, the etcd project's first priority is reliability.
Ensuring consistency and data safety requires our project contributors to continuously improve testing methodologies.
In this article, we will describe how we used advanced simulation testing to uncover subtle bugs,
validate the robustness of our releases, and increase our confidence in etcd's stability.
We'll share our key findings and how they have improved etcd.

## Enhancing etcd's Robustness Testing

Many critical software systems depend on etcd to be correct and consistent, most notably as the primary datastore for Kubernetes.
After some issues with the v3.5 release,
the etcd maintainers developed a new [robustness testing framework](https://github.com/etcd-io/etcd/issues/14045)
to better test for correctness under various failure scenarios. To further enhance our testing capabilities,

> we integrated an simulation testing platform from [Antithesis](https://antithesis.com/) into our workflow.

Reviewer 2: ```suggestion
we integrated a simulation testing platform from [Antithesis](https://antithesis.com/) into our workflow.
```

The platform works by running the entire etcd cluster inside a deterministic hypervisor.
This specialized environment gives the testing software complete control over every source of non-determinism,
such as network behavior, thread scheduling, and system clocks.
This means any bug it discovers can be perfectly and reliably reproduced.

Within this simulated environment, the testing methodology shifts away from traditional, scenario-based tests.
Instead of writing tests imperatively with strict assertions for one specific outcome, this approach uses declarative, property-based assertions about system behavior.
These properties are high-level invariants about the system that must always hold true. For example,
"data consistency is never violated" or "a watch event is never dropped."

The platform then treats these properties not as passive checks, but as targets to break.
It combines automated exploration with targeted fault injection,
actively searching for the precise sequence of events and failures that will cause a property to be violated.
This active search for violations is what allows the platform to uncover subtle bugs that result from complex combinations of factors.

> Antithesis refers to this approach as autonomous testing.

Reviewer 3: ```suggestion
Antithesis refers to this approach as "autonomous testing".
```

Reviewer 3: Should either be quoted or should be capitalized (the latter if it's used as a term of art or a trademark)

This builds upon etcd's existing robustness tests, which also use a property-based approach.
However, without a deterministic environment or automated exploration,
the original framework resembled throwing darts while blindfolded and hoping to hit the bullseye.
A bug might be found, but the process relies heavily on random chance and is difficult to reproduce.
Antithesis's deterministic simulation and active exploration
removes the blindfold, enabling a systematic and reproducible search for bugs.

Our goals for this testing effort were to:

1. **Validate the robustness of etcd v3.6.**
2. **Improve etcd's software quality by finding and fixing bugs.**
3. **Enhance our existing testing framework with autonomous testing.**

> ## How We Tested

Reviewer 4: Looks great to me.

We ran our existing robustness tests on the Antithesis simulation platform, testing a 3-node and a 1-node etcd cluster against a variety of faults, including:

* **Network faults:** latency, congestion, and partitions.

> * **Container-level faults:** node pauses, node kills, clock jitter, and CPU throttling.

Reviewer 5: does it support node-level freezing? or is it just pausing container

---
title: Autonomous Testing of etcd's Robustness

> author: "Marek Siarkowicz, Google & Craig Alfieri, Antithesis"

Author: DOne

draft: true
---

As a critical component of many production systems, including Kubernetes, the etcd project's first priority is reliability.
Ensuring consistency and data safety requires our project contributors to continuously improve testing methodologies.
In this article, we will describe how we used advanced simulation testing to uncover subtle bugs,
validate the robustness of our releases, and increase our confidence in etcd's stability.
We'll share our key findings and how they have improved etcd.

## Enhancing etcd's Robustness Testing

Many critical software systems depend on etcd to be correct and consistent, most notably as the primary datastore for Kubernetes.

> After some issues with the v3.5 release, the etcd maintainers developed a new After some issues with the v3.5 release,

Reviewer 6: ```suggestion
After some issues with the v3.5 release,
```

Looks like there's some leftover text on this line.

Author: Thanks, done

Within this simulated environment, the testing methodology shifts away from traditional, scenario-based tests.
Instead of writing tests with strict assertions that check for one specific outcome,
this approach relies on flexible, property-based assertions.
These properties are high-level invariants about the system that must always hold true. For example,
"data consistency is never violated" or "a watch event is never dropped."

The platform then treats these properties not as passive checks, but as targets to break.
It combines automated exploration with targeted fault injection,
actively searching for the precise sequence of events and failures that will cause a property to be violated.
This active search for violations is what allows the platform to uncover subtle bugs that result from complex combinations of factors.
Antithesis refers to this approach as autonomous testing.

This builds upon etcd's existing robustness tests, which also use a property-based approach.
However, without a deterministic environment or automated exploration,
the original framework resembled throwing darts while blindfolded and hoping to hit the bullseye.

> A bug might be found, but the process relies heavily on random chance and is difficult to reproduce.

Reviewer 3: @CraigAlfieri please review, thanks!

Reviewer 4: @jberkus Serdar (@sbenderli) and Akshay (@akshayjshah) are on the Antithesis Team and will be contributing thoughts here too.

The deterministic simulation and active exploration of the new platform,

> by contrast, effectively removes the blindfold, enabling a systematic and reproducible search for bugs.

Reviewer 7: I can understand the reproducibility. But can you explain why it is more likely to find bugs? Does Antithesis do the exploration by injecting random sequence of failures? If so, it is still a huge search space. Why is it more effective than robustness tests? Is it because it is automated? Or because we only inject one type of failure at a time in robustness tests?

Author: Are you asking to clarify the blogpost, or is it just curiosity? 

For me the main advantage is the [deterministic simulation testing](https://antithesis.com/resources/deterministic_simulation_testing/). It allows for guided space exploration with conscious failure injection, as opposed to random. Some advantages:
* Injecting failures around assertions to increase probability of finding failures.
* Controlled failure injection allowing to stack failures

[known-5-issue]: https://github.com/etcd-io/etcd/pull/13505
[known-5-report]: https://linuxfoundation.antithesis.com/report/6zkMqkEjjJuinArwLZTkJehM/7h42qHA5soBgxYPsMiDa4dSbA-O_g2SL9vkJqxvGON8.html?auth=v2.public.eyJzY29wZSI6eyJSZXBvcnRTY29wZVYxIjp7ImFzc2V0IjoiN2g0MnFIQTVzb0JneFlQc01pRGE0ZFNiQS1PX2cyU0w5dmtKcXh2R09OOC5odG1sIiwicmVwb3J0X2lkIjoiNnprTXFrRWpqSnVpbkFyd0xaVGtKZWhNIn19LCJuYmYiOiIyMDI1LTA1LTMwVDAwOjUyOjM2LjUwMjc3MTMyNFoifSYHsw1ZLCSfxME2keN58uGgi2yHTLvlg5_mFLkmePovjDjan-8SH72WdrmeWc4OMoRR-F3Pmi9UkU546_rtkgI#/run/8b82751143d9baaf98535089301d7af4-31-8/finding/984b7ce364030642155dcd71d492711c9f9f73a9

## Conclusion

The integration of this advanced simulation testing into our development workflow has been a success.
It has allowed us to find and fix critical bugs, improve our existing testing framework,

> and increase our confidence in the reliability of etcd. We will continue to leverage this technology

Reviewer 7: Are we going to run the Antithesis tests continuously or reevaluate for certain hours every certain period?

---
title: Autonomous Testing of etcd's Robustness
author: "Marek Siarkowicz, Google & Craig Alfieri, Antithesis"
date: 2025-08-13
draft: false
---

> Ensuring the reliability of etcd is our top priority. As a critical component of many production systems, we are continuously improving our testing methodologies. This blog post describes how we used autonomous testing to uncover subtle bugs, validate the robustness of our releases, and increase our confidence in etcd's stability. We'll share our key findings and how they have improved etcd.

Reviewer 4: This looks much tighter! Thank you Josh.

## Enhancing etcd's Robustness Testing

> Etcd is a critical component in many production systems, most notably as the primary datastore for Kubernetes. After some issues with the v3.5 release, the etcd maintainers developed a new [robustness testing framework](https://github.com/etcd-io/etcd/issues/14045) to better test for correctness under various failure scenarios. To further enhance our testing capabilities, we integrated an autonomous testing platform from [Antithesis](https://antithesis.com/) into our workflow. This allowed us to perform deterministic simulation testing for our distributed system.

Reviewer 3: ```suggestion
Other software systems depend on etcd to be correct and consistent, most notably as the primary datastore for Kubernetes. After some issues with the v3.5 release, the etcd maintainers developed a new [robustness testing framework](https://github.com/etcd-io/etcd/issues/14045) to better test for correctness under various failure scenarios. To further enhance our testing capabilities, we integrated an autonomous testing platform from [Antithesis](https://antithesis.com/) into our workflow. This allowed us to perform deterministic simulation testing for our distributed system.

Autonomous testing is a methodology for testing complete systems: end-to-end testing.  When using it, tests are automatically created by deterministic machine learning (AI/ML) automation in order to generate vastly more comprehensive test coverage than is feasible with just human-written tests. Antithesis' tools combine this with fault injection in order to reproduceably trigger bugs that might not be visible in simpler testing approaches.  Together, these approaches help the etcd project find issues that result from combinations of failures.
```

We need to define autonomous testing here.

Reviewer 3: We also want a link to a 3rd-party source that defines autonomous testing, which isn't Antithesis's website.

Reviewer 3: Craig, can you help with this?

Reviewer 3: Hey,  I really wanted to get Craig's feedback on this, thanks.

Reviewer 3: We need a summary that works for this article.  "It's not defined yet" isn't going to work; if it's not defined, then we shouldn't use the term.  We need a 2 to 3 sentence definition which is good enough for the blog article.

Reviewer 3: Oh, I see, you added a much longer explanation.  Lemme start out by reviewing that.

Reviewer 4: Hi Josh, this description works if we drop the first portion of the write-up...

Antithesis, uses the term Autonomous Testing to distinguish from "automated testing" or "test automation" -- when you use a piece of software to run tests that humans write. Antithesis defines autonomous testing as "having software write/design the tests for you."

Our goals for this testing effort were to:

> 1.  **Validate the robustness of etcd v3.6.**
> 2.  **Improve etcd's software quality by finding and fixing bugs.**
> 3.  **Enhance our existing testing framework with autonomous testing.**

Reviewer 8: The linter is complaining about the spaces after the bullet points or the numbers.

```suggestion
1. **Validate the robustness of etcd v3.6.**
2. **Improve etcd's software quality by finding and fixing bugs.**
3. **Enhance our existing testing framework with autonomous testing.**
```

## How We Tested

We ran our existing robustness tests on an autonomous testing platform, testing a 3-node and a 1-node etcd cluster against a variety of faults, including:

> *   **Network faults:** latency, congestion, and partitions.
> *   **Container-level faults:** node pauses, node kills, clock jitter, and CPU throttling.

Reviewer 8: Same here

```suggestion
* **Network faults:** latency, congestion, and partitions.
* **Container-level faults:** node pauses, node kills, clock jitter, and CPU throttling.
```

We tested older versions of etcd with known bugs to validate the testing methodology, as well as our stable releases (3.4, 3.5, 3.6) and the main development branch. In total, we ran 830 wall-clock hours of testing, which simulated 4.5 years of usage.

## What We Found

The results were impressive. The autonomous testing not only found all the known bugs we tested for but also uncovered several new issues in our main development branch.

Here are some of the key findings:

> *   **A critical watch bug was discovered** that our existing tests had missed. This bug was present in all stable releases of etcd.
> *   **All known bugs were found**, giving us confidence in the ability of the combined testing approach to find regressions.
> *   **Our own testing was improved** by revealing a flaw in our linearization checker model.

Reviewer 8: And here

```suggestion
* **A critical watch bug was discovered** that our existing tests had missed. This bug was present in all stable releases of etcd.
* **All known bugs were found**, giving us confidence in the ability of the combined testing approach to find regressions.
* **Our own testing was improved** by revealing a flaw in our linearization checker model.
```

> ### Issues in Main Development Branch

Reviewer 8: I think we're missing a "the" here

```suggestion
### Issues in the Main Development Branch
```

| **3** | Panic when two snapshots are received in short period | **Reported Issue Unable to be Reproduced, Found by Antithesis** | [Triage Report](https://linuxfoundation.antithesis.com/report/HqTiW-VhiXU25CCPP8vkSUPB/3Q73gnvlcEpEb6XVWcl4H3qTOnXZ7pFAdkpbpHr8mMI.html?auth=v2.public.eyJzY29wZSI6eyJSZXBvcnRTY29wZVYxIjp7ImFzc2V0IjoiM1E3M2dudmxjRXBFYjZYVldjbDRIM3FUT25YWjdwRkFka3BicEhyOG1NSS5odG1sIiwicmVwb3J0X2lkIjoiSHFUaVctVmhpWFUyNUNDUFA4dmtTVVBCIn19LCJuYmYiOiIyMDI1LTA2LTA2VDAxOjA5OjE5Ljc1MDg1NDI2NVoifW8RMYqVcS2V3idTzyvalEO2SnPqycds-Cn710lY-wlfqYPe1MAb2U0R2wEKVwPtSsr79WcnR8yYCyZyCQNqhAc#/run/31f74082d85b5ffdaf9f34ed37480bbd-32-6/finding/138fa550c81efa6efc7170191b75c4a22caea51f) | Open | Low | First found in May 2024 |

> | **4** | Panic from db page expected to be 5 but identifies as 0 | **Reported Issue Unable to be Reproduced, Found by Antithesis** | [Triage Report](https://linuxfoundation.antithesis.com/report/G-9rIjiZJiwodTEN5avQ7wgK/u_uFsWOwZSxS5mOmbEprwMUijNhsWdV6mfde_CT-y4k.html?auth=v2.public.eyJuYmYiOiIyMDI1LTA3LTAzVDA3OjMxOjUyLjQzMzQ2ODk3NFoiLCJzY29wZSI6eyJSZXBvcnRTY29wZVYxIjp7ImFzc2V0IjoidV91RnNXT3daU3hTNW1PbWJFcHJ3TVVpak5oc1dkVjZtZmRlX0NULXk0ay5odG1sIiwicmVwb3J0X2lkIjoiRy05cklqaVpKaXdvZFRFTjVhdlE3d2dLIn19fUHU0wnVRoDtfilwOCROUiDTtcOlIZkrVaddCqjorH3utgcIEPIzlsrMAJGXFC6NTZMneLqAWWU_lq-9prD_tQc#/run/9088730ba7972869a3e2b68b66708b55-32-6/finding/b9cbdf1bc8bd74cab1388e30ebdbf0b37c6f1420) | Open | Low | Previously reported; maintainers closed issue without fixing |

Reviewer 8: In these instances, where we state that we knew the issue or that it was already reported, would it make sense to add a link to that issue?

Author: Yes, thanks.
